code/espurna/ntp.ino

```diff
@@ -2,264 +2,343 @@
+    settimeofday(&tv, &tz);
 }
 
 // -----------------------------------------------------------------------------
 
 void ntpSetup() {
 
-    _ntpBackwards();
+    // Randomized in time to avoid clogging the server with simultaneous requests from multiple devices
+    // (for example, when multiple devices start up at the same time)
+    const uint32_t startup_delay = getSetting("ntpStartDelay", NTP_START_DELAY);
+    const uint32_t update_delay = getSetting("ntpUpdateIntvl", NTP_UPDATE_INTERVAL);
 
-    #if TERMINAL_SUPPORT
-        terminalRegisterCommand(F("NTP"), [](Embedis* e) {
-            if (ntpSynced()) {
-                _ntpReport();
-                terminalOK();
-            } else {
-                DEBUG_MSG_P(PSTR("[NTP] Not synced\n"));
-            }
-        });
+    _ntp_startup_delay = secureRandom(startup_delay, startup_delay * 2);
+    _ntp_update_delay = secureRandom(update_delay, update_delay * 2);
+    DEBUG_MSG_P(PSTR("[NTP] Startup delay: %us, Update delay: %us\n"),
+        _ntp_startup_delay, _ntp_update_delay
+    );
 
-        terminalRegisterCommand(F("NTP.SYNC"), [](Embedis* e) {
-            _ntpWantSync();
-            terminalOK();
-        });
-    #endif
+    _ntp_startup_delay = _ntp_startup_delay * 1000;
+    _ntp_update_delay = _ntp_update_delay * 1000;
 
-    NTPw.onNTPSyncEvent([](NTPSyncEvent_t error) {
-        if (error) {
-            if (error == noResponse) {
-                DEBUG_MSG_P(PSTR("[NTP] Error: NTP server not reachable\n"));
-            } else if (error == invalidAddress) {
-                DEBUG_MSG_P(PSTR("[NTP] Error: Invalid NTP server address\n"));
-            }
-            #if WEB_SUPPORT
-                wsPost(_ntpWebSocketOnData);
-            #endif
-        } else {
-            _ntp_report = true;
-            setTime(NTPw.getLastNTPSync());
-        }
-    });
+    // start up with some reasonable timestamp already available
+    _ntpSetTimestamp(__UNIX_TIMESTAMP__);
+
+    // will be called every time after ntp syncs AND loop() finishes
+    settimeofday_cb(_ntpSetTimeOfDayCallback);
 
-    wifiRegister([](justwifi_messages_t code, char * parameter) {
-        if (code == MESSAGE_CONNECTED) {
-            if (!ntpSynced()) {
-                _ntp_defer.once_ms(secureRandom(NTP_START_DELAY, NTP_START_DELAY * 15), _ntpWantSync);
-            }
+    // generic configuration, always handled
+    espurnaRegisterReload(_ntpConfigure);
+    _ntpConfigure();
+
+    // make sure our logic does know about the actual server
+    // in case dhcp sends out ntp settings
+    static WiFiEventHandler on_sta = WiFi.onStationModeGotIP([](WiFiEventStationModeGotIP) {
+        const auto server = _ntpGetServer();
+        if (sntp_enabled() && (!_ntp_server.length() || (server != _ntp_server))) {
+            DEBUG_MSG_P(PSTR("[NTP] Updating `ntpServer` setting from DHCP: %s\n"), server.c_str());
+            _ntp_server = server;
+            setSetting("ntpServer", server);
```

Comment:
Dhcp routine will call sntp_init() whenever dhcp server sets ntp option:
https://github.com/xoseperez/espurna/pull/1828
I also wonder if this is the reason that sntp can never be optimised away